proxysrv/server.go

```diff
@@ -0,0 +1,238 @@
+		defer cancel()
+	}
+	metrics := mlist.Metrics
+	span.Add(ssf.Count("import.metrics_total", float32(len(metrics)), map[string]string{
+		"veneurglobalonly": "",
+		"protocol":         "grpc",
+	}))
+
+	var errs []error
+
+	dests := make(map[string][]*metricpb.Metric)
+	for _, metric := range metrics {
+		dest, err := s.destForMetric(metric)
+		if err != nil {
+			errs = append(errs, s.recordError(span, err, "no-destination",
+				"failed to get a destination for a metric", 1))
+		} else {
+			// Lazily initialize keys in the map as necessary
+			if _, ok := dests[dest]; !ok {
+				dests[dest] = make([]*metricpb.Metric, 0, 1)
+			}
+			dests[dest] = append(dests[dest], metric)
+		}
+	}
+
+	// Wait for all of the forward to finish
+	wg := sync.WaitGroup{}
+	wg.Add(len(dests))
+
+	for dest, batch := range dests {
+		go func(dest string, batch []*metricpb.Metric) {
+			defer wg.Done()
+			if err := s.forward(ctx, dest, batch); err != nil {
+				errs = append(errs, s.recordError(span, err, "forward",
+					fmt.Sprintf("failed to forward %d metrics to the host '%s'",
+						len(batch), dest),
+					len(batch)))
+			}
+		}(dest, batch)
+	}
+
+	wg.Wait() // Wait for all the above goroutines to complete
+
+	protocolTags := map[string]string{"protocol": "grpc"}
+	span.Add(ssf.RandomlySample(0.1,
+		ssf.Timing("proxy.duration_ns", time.Since(span.Start), time.Nanosecond,
+			protocolTags),
+		ssf.Count("proxy.proxied_metrics_total", float32(len(metrics)), protocolTags),
+	)...)
+
+	s.opts.log.WithFields(logrus.Fields{
+		"protocol": "grpc",
+		"duration": time.Since(span.Start),
+	}).Info("Completed forwarding to downstream Veneurs")
+
+	return errSliceToErr(errs)
+}
+
+// recordError records when an error has resulted in some metrics not being
+// forwarded.  It submits diagnostic metrics, logs an error, and then returns
+// a wrapped error.
+func (s *Server) recordError(
+	span *trace.Span,
+	err error,
+	cause string,
+	message string,
+	numMetrics int,
+) error {
+	tags := map[string]string{
+		"cause":    cause,
+		"protocol": "grpc",
+	}
+	span.Add(ssf.Count("proxy.proxied_metrics_failed", float32(numMetrics), tags))
+	span.Add(ssf.Count("proxy.forward_errors", 1, tags))
+	s.opts.log.WithError(err).WithFields(logrus.Fields{
+		"cause": cause,
+	}).Error(message)
+
+	return fmt.Errorf("%s: %v", message, err)
+}
+
+// destForMetric returns a destination for the input metric.
+func (s *Server) destForMetric(m *metricpb.Metric) (string, error) {
+	key := samplers.NewMetricKeyFromMetric(m)
+	dest, err := s.destinations.Get(key.String())
+	if err != nil {
+		return "", fmt.Errorf("failed to hash the MetricKey '%s' to a "+
+			"destination: %v", key.String(), err)
+	}
+
+	return dest, nil
+}
+
+// forward sends a set of metrics to the destination address, and returns
+// an error if necessary.
+func (s *Server) forward(ctx context.Context, dest string, ms []*metricpb.Metric) (err error) {
+	conn, err := grpc.Dial(dest, grpc.WithInsecure())
```

Comment:
Same here as above - this should be done up front, not on the send path, i think.